pkg/log/logger.go

```diff
@@ -0,0 +1,202 @@
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+// Logger The global logger
+var Logger logr.Logger
+
+// FromContextOrGlobal return a logger from the passed context or the
+// global logger.
+func FromContextOrGlobal(ctx context.Context) logr.Logger {
+	if logger, err := logr.FromContext(ctx); err == nil {
+		return logger
+	}
+	return Logger
+}
+
+// New returns a new logger. The level may be "DEBUG" (V(1)) or "TRACE" (V(2)),
+// any other string (e.g. "") is interpreted as "INFO" (V(0)). On first call
+// the global Logger is set.
+func New(name, level string) logr.Logger {
+	logger := newLogger(level).WithName(name)
+	once.Do(func() { Logger = logger })
+	return logger
+}
+
+var once sync.Once
+
+// Fatal log the message using the passed logger and terminate
+func Fatal(logger logr.Logger, msg string, keysAndValues ...interface{}) {
+	if z := zapLogger(logger); z != nil {
+		z.Sugar().Fatalw(msg, keysAndValues...)
+	} else {
+		// Fallback to go default
+		golog.Fatal(msg, keysAndValues)
+	}
+}
+
+// NSMLogger return a logger to use for NSM logging.
+func NSMLogger(logger logr.Logger) nsmlog.Logger {
+	// Get the zap logger
+	z := zapLogger(logger)
+	if z == nil {
+		panic("NSMLogger: Can't get the Zap logger")
+	}
+	nsmz := z.With(zap.String("subsystem", "NSM"))
+	return &nsmLogger{
+		z: nsmz,
+		s: nsmz.Sugar(),
+	}
+}
+
+// Called before "main()". Pre-set a global logger
+func init() {
+	Logger = newLogger("").WithName("Meridio")
+}
+
+func timeEncoder(t time.Time, enc zapcore.PrimitiveArrayEncoder) {
+	enc.AppendString(t.Format("2006-01-02T15:04:05.999-07:00"))
+}
+
+func levelEncoder(lvl zapcore.Level, enc zapcore.PrimitiveArrayEncoder) {
+	switch lvl {
+	case zapcore.InfoLevel:
+		enc.AppendString("info")
+	case zapcore.WarnLevel:
+		enc.AppendString("warning")
+	case zapcore.ErrorLevel:
+		enc.AppendString("error")
+	case zapcore.DPanicLevel:
+		enc.AppendString("critical")
+	case zapcore.PanicLevel:
+		enc.AppendString("critical")
+	case zapcore.FatalLevel:
+		enc.AppendString("critical")
+	default:
+		enc.AppendString("debug")
+	}
+}
+
+func newLogger(level string) logr.Logger {
+	var lvl int
+	switch level {
+	case "DEBUG":
+		lvl = -1
+	case "TRACE":
+		lvl = -2
+	default:
+		lvl = 0
+	}
+	zc := zap.NewProductionConfig()
+	zc.Level = zap.NewAtomicLevelAt(zapcore.Level(lvl))
+	zc.DisableStacktrace = true
+	zc.DisableCaller = true
+	zc.EncoderConfig.NameKey = "service_id"
+	zc.EncoderConfig.LevelKey = "severity"
+	zc.EncoderConfig.TimeKey = "timestamp"
+	zc.EncoderConfig.MessageKey = "message"
+	//zc.EncoderConfig.EncodeTime = zapcore.ISO8601TimeEncoder (almost works)
+	zc.EncoderConfig.EncodeTime = timeEncoder
+	zc.EncoderConfig.EncodeLevel = levelEncoder
+	zc.Encoding = "json"
+	z, err := zc.Build()
+	if err != nil {
+		panic(fmt.Sprintf("Can't create a zap logger (%v)?", err))
+	}
+	return zapr.NewLogger(z.With(
+		zap.String("version", "1.0.0"), zap.Namespace("extra_data")))
```

Comment:
Is this supposed to be the Meridio Version?

Reply:
No, it's the version of the ADP format. It is _required_ according to the ADP format spec.

Reply:
```json
  "required": [
    "version",
    "timestamp",
    "severity",
    "service_id",
    "message"
  ]
```